### PR TITLE
Add API gateway

### LINKED ISSUE #14: F2.1.1 The service should expose an endpoint for the admin to create new questions and include the 5 compulsory attributes (Id, Title, Description, Category, Complexity).

### PR DESCRIPTION
**Pull request by jq1836**
Resolves #14, #19, #20, #24 

## Overview
Previously, our question service was exposed, allowing anyone to use the endpoints. The API gateway provides automatic authentication, as well as provides ingress capabilities.

Currently does not provide load balancing capabilities. Intention is for load balancing to be done by **k8s**.

## Testing instructions
1. Add in `API_GATEWAY_PORT` value to `.env`. Note that whatever port you set here will determine how you access the pages.
2. docker compose up --build.
3. Go to `localhost:$API_GATEWAY_PORT` to access frontend.
4. Play around with backend endpoints through frontend.
5. Play around with backend endpoints through preferred requester, port is always $API_GATEWAY_PORT.

### CODE REVIEW

**Comment by jq1836** [comment]
To review after PR #166

**Review by Simongolovinskiy — APPROVED** [review]

File: frontend/lib/update-user-privilege.ts [hunk 1]
```diff
@@ -8,7 +8,7 @@ export const updateUserPrivilege = async (
   const body = { isAdmin };
 
   const response = await fetch(
-    `${userServiceUri(window.location.hostname)}/users/${id}/privilege`,
+    `${userServiceUri(window.location.hostname, AuthType.Public)}/users/${id}/privilege`,
```
Comment (SelwynAng) [inline frontend/lib/update-user-privilege.ts:11]: Not sure if I understand correctly, but should this be `AuthType.Public`? Updating user privilege should be an admin function?

File: frontend/lib/update-user.ts [hunk 2]
```diff
@@ -15,7 +15,7 @@ export const updateUser = async (
   const body = { username, email, password, skillLevel };
 
   const response = await fetch(
-    `${userServiceUri(window.location.hostname)}/users/${id}`,
+    `${userServiceUri(window.location.hostname, AuthType.Public)}/users/${id}`,
```
Comment (SelwynAng) [inline frontend/lib/update-user.ts:18]: Same as above, since updating user should be done by logged in users only?

**Review by SelwynAng — COMMENTED** [review]
LGTM overall, aside from some comments. 

Also, should we set the API gateway port to some fixed value instead of making it flexible? Tested with `8080` and `2000`, those port numbers don't allow the frontend to communicate with the backend services. Port numbers like `80` and `1234` work though.

**Review by wr1159 — APPROVED** [review]
LGTM - pz

**Review by SelwynAng — APPROVED** [review]
LGTM!

**Pull request merged**